bdb-ip.md

```diff
@@ -0,0 +1,437 @@
+# Improvement proposals for HTTP API
+
+
+Currently, there is two milestones grouping issues for upcoming HTTP API
+endpoints:
+
+- [Changes that break the existing HTTP API of BDB
+  v0.11](https://github.com/bigchaindb/bigchaindb/milestone/11)
+- [Changes that are additions to the exiting HTTP API of BDB
+  v0.11](https://github.com/bigchaindb/bigchaindb/milestone/9)
+
+After releasing BDB v1.0, breaking changes to the external APIs are not
+permitted anymore. Additions to the APIs are however always allowed.
+
+To resolve the outstanding issues, we'll hence limit
+the scope of this document to the already existing endpoints.
+
+
+### First task
+
+Please check the two listed milestones above for the correctness of categorized
+issues.
+
+
+### Methodology
+
+This document will go address each of the issues in the [breaking
+changes milestone](https://github.com/bigchaindb/bigchaindb/milestone/11) and
+discuss possible solutions to them. Once there is a consensus on the solutions,
+we can implement them into BDB core (mainly HTTP API), py and js driver and
+the BDB CLI.
+
+
+### Issues in milestone
+
+#### [Inconsistent naming of short-hand ids](https://github.com/bigchaindb/bigchaindb/issues/1134)
+
+Current naming of ids according to [HTTP API documentation
+(master)](https://docs.bigchaindb.com/projects/server/en/master/http-client-server-api.html)
+and the [Websocket Event Stream API
+(master)](https://docs.bigchaindb.com/projects/server/en/master/websocket-event-stream-api.html):
+
+- Transaction id:
+    - `tx_id` in query parameters and Event Stream API
+    - `txid` in `transaction.inputs.fulfills`
+    - `id` in transaction payload
+    - `id` in `/assets?search={text_search}` (this is also the asset's id!)
+    - `ID` in prose text
+- Asset id:
+    - `asset_id` in query parameters and Event Stream API
+    - `id` in `transaction.asset` payload
+    - `id` in `/assets?search={text_search}` (this is also the transaction's id!)
+- Fulfillment id:
+    - `type_id` in `transaction.outputs.condition.details`
+- Block id:
+    - `block_id` in query parameters and Event Stream API
+    - `id` in the block payload
+
+General inconsistency: `block_id`, `asset_id` and `tx_id` instead of
+`transaction_id`.
+
+
+Even though, this shouldn't be an argument (as we see the HTTP API descoped
+from the underlying implementation):
+
+- [`txid` usage in bdb-server](https://docs.bigchaindb.com/projects/server/en/master/websocket-event-stream-api.html)
+- [`txid` usage in bdb-server](https://docs.bigchaindb.com/projects/server/en/master/websocket-event-stream-api.html)
+
+
+##### Proposal 1
+
+- Call any identifier that points to an object externally is a `object_id`
+    - e.g. `transaction.inputs.fulfills.tx_id` points to an external tx, hence
```

Comment:
If an identifier within the data is `object.id` and an identifier that points to an external object is `object_id` shouldn't `tx_id` actually be `transaction_id`?